index.js

```diff
@@ -855,19 +855,24 @@ function hexWrite (buf, string, offset, length) {
 
   const strLen = string.length
 
-  if (length > strLen / 2) {
-    length = strLen / 2
+  if (length > (strLen >>> 1)) {
+    length = strLen >>> 1
   }
-  let i
-  for (i = 0; i < length; ++i) {
-    const a = hexCharValueTable[string[i * 2]]
-    const b = hexCharValueTable[string[i * 2 + 1]]
-    if (a === undefined || b === undefined) {
+
+  for (let i = 0; i < length; ++i) {
+    const a = string.charCodeAt(i * 2 + 0)
+    const b = string.charCodeAt(i * 2 + 1)
+    const hi = hexCharValueTable[a & 0x7f]
+    const lo = hexCharValueTable[b & 0x7f]
+
+    if ((a | b | hi | lo) & ~0x7f) {
```

Comment:
This line is kind of magic: it's checking for multiple things. If `a` or `b` exceed `0x7f` (ascii), the check will fail. If `a` or `b` result in a lookup of a non-hex character in `hexCharValueTable`, `hi` and/or `lo` become `-1`, also causing the check to fail.

This is the only I know of to handle all possible errors in a single branch. I used this same trick in [bcrypto][1]. It seems to work well.

[1]: https://github.com/bcoin-org/bcrypto/blob/2b9f954/lib/js/base16.js#L94

Reply:
Also, `a` and `b` are ANDed by `0x7f` before the array access to prevent the function from being deoptimized due to an OOB access.

Reply:
Some examples:

```
> Buffer.from('ab\u2015c', 'hex')
<Buffer ab>
> Buffer.from('ab\u1520c', 'hex')
<Buffer ab>
> Buffer.from('ab\u0715c', 'hex')
<Buffer ab>
```

Reply:
Maybe an easy way to assure this could be to add some tests?

Reply:
Tests written. It tests the entire 16 bit space along with another test for odd-length strings.

Reply:
Ha, that's brilliant.  Thanks @chjj